### ISSUE TITLE
Variable field drop downs contain typed variables

### ISSUE THREAD

**Issue by NAllred91** [google/blockly#1921]
### Problem statement

If you drag a variable block (either set/change/get) into the workspace, and then click the variable field dropdown, you will see that typed and untyped variables both show up in the dropdown.

### Expected Behavior

The previous behavior was that the variable fields on the variable blocks (set/change/get) only included untyped variables in the drop down.

I haven't been able to find what change in Blockly has caused this.  But as far as I can tell I wasn't doing anything on my end to cause this behavior.

**Comment by AnmAtAnm** [comment]
I believe this is works as intended.

Given that the variable dropdown shows all types of variables, even if the variable block is attached (setter's child or getter's parent), it makes sense to assume it would show untyped variables. Even assuming such connected block type constraints, an untyped variable would work for all blocks, since it behaves as an "any" type.

Using both typed and untyped variables in the same workspace seems like an unusual case. Can you elaborate on how you are using them together.

**Comment by AnmAtAnm** [comment]
I just verified this behavior has not changed since the Februrary release, which I think is the first release we included the untyped variables in out playground.

**Comment by NAllred91** [comment]
I think I've found when this behavior was changed.

https://github.com/google/blockly/pull/1481/files#diff-fb72a00279b943f530ea656a48336806R142

core/field_variable.js line 142.

It appears that previously you had to explicitly set `opt_variableTypes` to `null` if you wanted all variables included in the drop down.  This pull request made a change in `Blockly.FieldVariable.prototype.getVariableTypes_` so that if variableTypes is null _or_ undefined, return all variable types.

Eventually, @rachel-fenichel refactored some of this and moved the logic to `setTypes_` (where it exists now), but it appears that this is the first place that `if (variableTypes === null) {` was changed to `if (variableTypes === null || variableTypes === undefined) {`


Prior to this PR we had this.  The explicit check against null ensured that if `opt_variableTypes` was not defined, then only untyped variables (or rather, variables of type `''`) would be shown in the dropdown.
```
Blockly.FieldVariable.prototype.getVariableTypes_ = function() {
  var variableTypes = this.variableTypes;
  if (variableTypes === null) {
    // If variableTypes is null, return all variable types.
    if (this.sourceBlock_) {
      var workspace = this.sourceBlock_.workspace;
      return workspace.getVariableTypes();
    }
  }
  variableTypes = variableTypes || [''];
  if (variableTypes.length == 0) {
    // Throw an error if variableTypes is an empty list.
    var name = this.getText();
    throw new Error('\'variableTypes\' of field variable ' +
      name + ' was an empty list');
  }
  return variableTypes;
};
```

I know this behavior that I ended up depending on only existed in Blockly for a short time, so I doubt you'll be interested in going back to that behavior.  But I'm wondering if it would make sense to make this behavior available again without the need for messing with Blockly source code?

**Comment by AnmAtAnm** [comment]
Thanks for the investigation and the writeup. 

Based on the description above, you are not using the typed variables as real variables. Your filter is one approach, but you might consider using a custom "device" dropdown field as another possibility. That said, you were also looking to ensure your names were unique with respect to the user variables, which does make the variable overload make some sort of sense.

**Comment by NAllred91** [comment]
I'm not sure I explained it very well.  I am using a custom "device" dropdown in order to let the user select a device.  For example, if a user adds two Sphero's, I add two "sphero" type variables, and then I get variables by type in order to populate the drop down in Sphero blocks.

They are real variables in the sense that when I output the javascript code the "sphero" typed variable is defined at the top of the code, and then I use that variable throughout the code.  The purpose is to keep these typed variables separate from any variables the user might want to create in their program.  For example, I don't want to give the user the the ability to modify these typed variables because that will break their code.

Its also really convenient because Blockly handles modifying the variable name if the user wants to call the device "var", or some other reserved word.

But my issue here is that when a user goes into the "variable" flyout (toolbox xml category with `custom="VARIABLE"`) and clicks "create variable...", when they then pull one of those variables into the workspace and click on the variable field drop down they see the typed variables.  I think that the user should _not_ see typed variables in this case and only see untyped variables.  Or maybe there should be a way to indicate if you want the drop down to include typed and untyped?

**Comment by NAllred91** [comment]
If it would be helpful I could put together a working example that demonstrates this?  I think it is possible that the behavior that I'm wanting to go back to could be useful outside of my specific use case.

**Comment by RoboErikG** [comment]
Adding a note to document the reasoning:
We made the change to use null || undefined so that the default behavior for a developer would be for variables of all types to be shown in the standard setter/getter for variables. Typed variables weren't fully fleshed out at the time and we decided it would be less surprising to see all variables show up than to see only some variables show up.

The recommended way to only show variables with the empty type is to add "variableTypes": [""] to the default getter and setter. The [variables docs](https://developers.google.com/blockly/guides/create-custom-blocks/variables) have more details.
